x/wasm/internal/keeper/options.go

```diff
@@ -1,28 +1,73 @@
 package keeper
 
-import "github.com/CosmWasm/wasmd/x/wasm/internal/types"
+import (
+	"fmt"
+	"github.com/CosmWasm/wasmd/x/wasm/internal/types"
+)
 
 type optsFn func(*Keeper)
 
 func (f optsFn) apply(keeper *Keeper) {
 	f(keeper)
 }
 
-// WithMessageHandler is an optional constructor parameter to replace the default wasm vm engine with the
+// WithMessageHandler is an optional constructor parameter to replace the default wasmVM engine with the
 // given one.
 func WithWasmEngine(x types.WasmerEngine) Option {
 	return optsFn(func(k *Keeper) {
-		k.wasmer = x
+		k.wasmVM = x
 	})
 }
 
-// WithMessageHandler is an optional constructor parameter to set a custom message handler.
+// WithMessageHandler is an optional constructor parameter to set a custom handler for wasmVM messages.
 func WithMessageHandler(x messenger) Option {
 	return optsFn(func(k *Keeper) {
 		k.messenger = x
 	})
 }
 
+// WithQueryHandler is an optional constructor parameter to set custom query handler for wasmVM requests.
+func WithQueryHandler(x wasmVMQueryHandler) Option {
+	return optsFn(func(k *Keeper) {
+		k.wasmVMQueryHandler = x
+	})
+}
+
+// WithQueryPlugins is an optional constructor parameter to pass custom query plugins for wasmVM requests.
+func WithQueryPlugins(x *QueryPlugins) Option {
+	return optsFn(func(k *Keeper) {
+		q, ok := k.wasmVMQueryHandler.(interface {
+			Merge(o *QueryPlugins) QueryPlugins
+		})
+		if !ok {
+			panic(fmt.Sprintf("Unsupported query handler type: %T", k.wasmVMQueryHandler))
+		}
+		k.wasmVMQueryHandler = q.Merge(x)
+	})
+}
+
+// WithMessageEncoders is an optional constructor parameter to pass custom message encoder to the default wasm message handler.
+func WithMessageEncoders(x *MessageEncoders) Option {
+	return optsFn(func(k *Keeper) {
+		q, ok := k.messenger.(*MessageHandlerChain)
+		if !ok {
+			panic(fmt.Sprintf("Unsupported message handler type: %T", k.messenger))
+		}
+		s, ok := q.handlers[0].(SDKMessageHandler)
+		if !ok {
+			panic(fmt.Sprintf("Unexpected message handler type: %T", q.handlers[0]))
+		}
+		e, ok := s.encoders.(interface {
```

Comment:
Maybe we just cast this to the concrete type we set up in `NewDefaultMessageHandler`. This is after 2x casting interfaces to concrete types, so if some funky modifications on the types happened already, you have no guarantees.